### PR TITLE
Add missing features, polish to `SynchronousComputeService`

### LINKED ISSUE #34: Compute : `SynchronousComputeService` implementation with core features
We require a compute service that serves as the reference implementation for any compute service that interacts with the `FahAlchemyComputeAPI` to pull tasks, execute `ProtocolDAG`s, and push results back. This reference implementation should perform all activity in-process, in-serial, for ease of testing and debugging.

More performance-focused compute services will build on the components implemented in this reference implementation.

### PR DESCRIPTION
**Pull request by dotsdl**
Closes #34.

### CODE REVIEW

File: alchemiscale/compute/service.py
Comment (hmacdope) [inline alchemiscale/compute/service.py:262]: I would also dump the scoped keys to the log?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few little queries but overall looking good.

**Comment by dotsdl** [comment]
@mikemhenry do you have a favorite logging formatter you'd like us to use for compute service logs?

File: alchemiscale/compute/service.py
Comment (dotsdl) [reply alchemiscale/compute/service.py:315]: We already indicate which task we execute and finish further below. I think the log may be too noisy doing it here as well.

File: alchemiscale/compute/api.py
Comment (dotsdl) [reply alchemiscale/compute/api.py]: Done!

**Comment by dotsdl** [comment]
Ready for final review @hmacdope!

File: alchemiscale/compute/api.py
Comment (hmacdope) [inline alchemiscale/compute/api.py:140]: Query: is there any need for this to return or expose the expired `compute_service_id`s? I'll leave it up to you, im sure this is fine for now.

File: alchemiscale/compute/service.py
Comment (hmacdope) [inline alchemiscale/compute/service.py:201]: nitpick: th clearer than tq
```suggestion
                list(taskhubs.keys()), weights=[th.weight for th in taskhubs.values()]
```

File: alchemiscale/compute/service.py
Comment (hmacdope) [inline alchemiscale/compute/service.py:217]: Im not quite sure I understand this pop, what if we didn't take all the tasks from the hub? I'm probably missing something.

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:123]: This looks good!

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:764]: Does this merge allow you to attempt to register the same `compute_service` more than once? IMO should be an error if you try and overwrite and existing service or at least give some feedback if the query did nothing ie found an already existing service. I could be missing something that prevents this though...

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:789]: nitpick but should return `n` from the cypher query and propagate that value if possible.

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:816]: Cypher does datetime comparison? Cool!

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:1312]: useful check, I'm sure there are other sharp edges of this nature floating around.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (hmacdope) [inline alchemiscale/tests/integration/storage/test_statestore.py:301]: test return values from database queries here also ?

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (hmacdope) [inline alchemiscale/tests/integration/storage/test_statestore.py:321]: Also check return values here?

File: alchemiscale/tests/integration/test_cli.py
Comment (hmacdope) [inline alchemiscale/tests/integration/test_cli.py:152]: If this YAML is required perhaps in example sometwhere in `devtools` is a good idea?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great @dotsdl, just a few minor things. Ping me again when you want another review, its super close.

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:123]: Thank you!

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:764]: Good question!

We're using `py2neo`'s `Transaction.merge` here: https://py2neo.org/2021.1/workflow.html?highlight=merge#py2neo.Transaction.merge

It won't create a `ComputeServiceRegistration` with the same `identifier`, and for additional safety on this we also set a uniqueness constraint for `ComputeServiceRegistration`s on their `identifier` (see above). That being said, no feedback will be given if the same `ComputeServiceRegistration` is trying to be used twice; I don't think this will be a problem in practice, but you raise a good point that we also don't want this happening.

I can change this to use `tx.create` instead, which will raise an exception if the object exists already due to the constraint we have set above.

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:789]: I don't think we can return `n` itself, since we are deleting it. However, I can use the trick we're using below to return the identifier after extracting it with `WITH` prior to the delete.

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:816]: It does! :grin:

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:1312]: There most certainly are; I tripped on this one myself while doing live testing. I'll sweep through the other methods that are susceptible to this problem and add appropriate checks.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (dotsdl) [reply alchemiscale/tests/integration/storage/test_statestore.py:301]: Ah, good catch; will add this.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (dotsdl) [reply alchemiscale/tests/integration/storage/test_statestore.py:321]: Will do!

File: alchemiscale/tests/integration/test_cli.py
Comment (dotsdl) [reply alchemiscale/tests/integration/test_cli.py:152]: Good call, will add an example file. I think this may get better with #107 as well.

File: alchemiscale/compute/api.py
Comment (dotsdl) [reply alchemiscale/compute/api.py:140]: Not at the moment no. The approach we take here (stacking in expiration with the heartbeat API call) is actually a bit of an inefficient hack; the better approach to this will be to have a separate `PeriodicsService` that performs operations like this on some schedule, since they only have to happen once in a while. This will avoid thrashing the database.

I'll create an issue articulating the vision for this service so we can implement for a future release.

File: alchemiscale/compute/service.py
Comment (dotsdl) [reply alchemiscale/compute/service.py:201]: Ah, thank you! Yes this is a leftover from the old `TaskQueue` implementation.

File: alchemiscale/compute/service.py
Comment (dotsdl) [reply alchemiscale/compute/service.py:217]: The idea I've tried to implement here is explained in the comments. I'll try to express it in a different way:
1. When we enter this method, we want to claim `count` `Task`s. 
2. We start by randomly selecting a `TaskHub` to draw from based on their relative weights.
3. We try to draw `count` `Task`s from that `TaskHub`. However, it's possible that this `TaskHub` doesn't actually have enough `Task`s to meet our desired `count`. So, we drop it from the set we select from and roll the dice again, repeating from (2).
4. We keep going until we have either collected `count` `Task`s or run out of `TaskHub`s to draw `Task`s from.

Does the implementation above accomplish this, in your view?

File: alchemiscale/compute/api.py
Comment (dotsdl) [reply alchemiscale/compute/api.py:140]: Issue here: https://github.com/openforcefield/alchemiscale/issues/109

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:1312]: I think I've filled many of the similar potholes in.

**Comment by dotsdl** [comment]
Changes complete @hmacdope! Ready for another review!

**Review by hmacdope — APPROVED** [review]
You satisfied all my queries @dotsdl! Feel free to merge when ready.

**Comment by dotsdl** [comment]
Thanks @hmacdope! This PR is much better following your reviews. Merging!

**Pull request merged**
